### ISSUE TITLE
Sharing Findings Regarding HUD Fix

### ISSUE THREAD

**Issue by wagrenier** [Lyall/MetaphorFix#21]
I have been following your work for a long while and I would be very glad to help you out with the HUD issues. Last week, I pulled the repo and tried a few things myself. The main issue I encountered was the health bar not showing the right value. I found that the red bar seems to be hidden unless overlayed with the actual health bar. By offsetting the `hpmp_gauge` by roughly 100 pixels the lost health bar reappeared.

Let me know if you have any further information that could help me with that, I'll continue researching if I find anything worthwhile.

Good luck on your future projects :)

**Comment by Lyall** [comment]
This is helpful thanks. I think the issue is with the approach, at the moment I think the edits are happening too early and this is causing a cascading series of issues further down the chain of HUD processing. In that manner, I actually ended up gutting all the HUD fix code and rethinking things. I've added some WIP HUD work to the debug build, if you want to play around with it and see, just switch to the debug build and compile it.

So far, this new approach seems promising. There's no values being adjusted mid-load or anything like what was happening before. That was the part that was really flummoxing me before and I think it stemmed from changing things too early on. So far the major issue with this approach is that there is some clipping going on. I suspect that is probably a shared function so if I can find that and account for the HUD offset, I think it might solve that.

As you can see in this shot, there's way less overall issues (other than the clipping).

![METAPHOR_2024_10_10_19_26_16_180](https://github.com/user-attachments/assets/d926326a-24b6-4777-96c6-5cdb92fa00e3)

**Comment by wagrenier** [comment]
Thanks a lot for the info, I'll take a look at that build tonight!

Indeed, I see the clipping on 2D world elements like the sand effects. Have you tried using a graphics debugger like RenderDoc? It could be helpful in figuring out how the HUD and 2D elements are drawn with what parameters. If not, I can try it out.

If you have any general function offsets for IDA/Ghidra it would definitely help in finding out the functions that cause the issues.

Finally, if you have a public discord or something like to more easily share info I'll gladly join. I can even help for future projects :)

No worries about the salad ;)

**Comment by Lyall** [comment]
Yeah I've used NSight to actually address the shadow resolution in the game. It was pretty helpful for finding out that the CSMs actually use a texture position offset that's hardcoded for 2048x2048 shadowmaps. So if you bump up the resolution of the shadowmaps, they look offset and blurry. I haven't found it too useful for 2D stuff in this game. In fact it seems like most of the UI rendering happens bit-by-bit instead of in batches, which seems untypical based on other games I've looked at.

I don't have any stuff mapped out in Ghidra or IDA. I've kind of skipped over using them since it's easier to rapidly prototype stuff with Cheat Engine and it's powerful auto-assembler. That said, once the full game is out, it probably would be a good idea to start working on an IDA database for the game. I have made note of some useful functions at least. I think a few of the ones I looked at previously for adjusting the HUD will prove useful later on when it comes to more granular HUD edits. Like I've seen people requesting to have the vignette at the edges of some elements removed. 

I don't have my own Discord server, though I have thought about starting one! I do tend to hang about on the [WSGF Discord](https://discord.com/invite/wsgf-142072549537349632) though, so that's where I'm usually found. 

I went back by the way and I was able to find where the HUD was being clipped. I'll push that commit in a bit once I polish it up. It's looking quite a bit better now. 

![METAPHOR_2024_10_10_20_35_34_378](https://github.com/user-attachments/assets/9c2a0816-c3f4-4b10-9776-8396968f4987)
